src/Nethermind/Nethermind.Consensus/Processing/ReadOnlyTxProcessingScope.cs

```diff
@@ -9,20 +10,31 @@ namespace Nethermind.Consensus.Processing;
 
 public class ReadOnlyTxProcessingScope(
     ITransactionProcessor transactionProcessor,
-    IWorldState worldState,
-    Hash256 originalStateRoot
+    IWorldState worldState
 ) : IReadOnlyTxProcessingScope
 {
+    private IDisposable? _worldStateScopeGuard;
+    private Hash256? _originalStateRoot = worldState.StateRoot;
     public void Dispose()
     {
         Reset();
+        _worldStateScopeGuard?.Dispose();
+    }
+
+    public void Init(Hash256 stateRoot)
+    {
+        _worldStateScopeGuard = worldState.BeginScope(stateRoot);
     }
 
     public ITransactionProcessor TransactionProcessor => transactionProcessor;
     public IWorldState WorldState => worldState;
     public void Reset()
     {
-        worldState.StateRoot = originalStateRoot;
-        worldState.Reset();
+        // this means the scope was not initialized here, so don't need to reset anything
+        if (_worldStateScopeGuard is not null)
+        {
+            worldState.StateRoot = _originalStateRoot;
```

Comment:
And here's the missing rootHash capturing I believe. If the `_worldStateScopeGuard ` captured the previous hash that it resets to on `.Dispose`, you'd need no manual capturing of the original state root.